### DISCUSSION TITLE
SQL Trigger : Statistics.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#288] [Q&A] [answered]
As I find the automatism of the "triggers" great, I open the discussion :

I explain the situation:
1 - With the rules of conservation, the history of the borrowed documents are erased after 45 days.
2 - We have to provide annual statistics, therefore on 365 days.
3 - The books lent by the central library do not remain in the database after 11 months.

Is there a way with the Triggers to create a new table that would keep the information anonymously?

desired informations:
- myoid (to have the number of documents)
- borrowing date (years)
- origin (BDP | Own funds)
- target_audience (Adults | Youth)
- title (most read book statistic)
- author (most read author statistics)

**Comment by meteos77** [comment]
```
CREATE TABLE statistics_history (
    id INTEGER PRIMARY KEY,
    new_myoid int,
    old_reserved_date text,
    old_originality text,
    old_target_audience text,
    old_title text,
    old_author text
);

CREATE TRIGGER statistics_history_trigger AFTER INSERT ON item_borrower
FOR EACH row								
BEGIN									
    INSERT INTO statistics_history (
        new_myoid
        old_reserved_date,
        old_originality,
        old_target_audience,
        old_title,
        old_author
    )
VALUES
    (
        new.myoid
        old.reserved_date,
        old.originality,
        old.target_audience,
        old.title,
        old.author
        
    )
END;	
```

**Reply by meteos77** [reply]
it seems to me too simple for a good functioning.

 Thanks to you the specialist for your comments / advice

**Comment by textbrowser** [comment]
BQ is not aware of local / state / international / galactic laws.

**Reply by meteos77** [reply]
BQ does not know but the small SQL query of Sunday is great because it allows to be within the limits of the regulation.
That's the main thing.

**Comment by textbrowser** [comment]
SQLite is a private database. Managing data is the administrator's responsibility. Create special DELETE queries and execute them as necessary.

Executing destructive queries which run in the dark is not wise.

**Comment by textbrowser** [comment]
BQ is not a daemon-like project. It doesn't perform periodic queries. It has all / most of the features for performing queries for administering data / patrons. SQLite doesn't have patrons. One or more people administer the same database. Anyone managing the data is an administrator. As you're migrating away from that through such requests, PQ is the optimal solution. PQ allows for periodic analysis.

http://www.postgresonline.com/journal/archives/19-Setting-up-PgAgent-and-Doing-Scheduled-Backups.html

In conclusion, I'm not adding such features in BQ.

**Reply by meteos77** [reply]
Thanks for all the clarifications.

I know that PostgreSQL would be better but I am not yet competent on the domain. ( probably project of the year 2024 )
PostgreSQL - for BQ base and OpenStreetMap base)

For 2023 :
- It is already necessary to switch to BiblioteQ the current base.
- Train the people to BiblioteQ.
I think at least until December 2023.

**Comment by textbrowser** [comment, marked as answer]
I'll enable optional history for SQLite too. Like PQ.
